File: epi_judge_python_solutions/max_product_all_but_one.py
import itertools
import operator

from test_framework import generic_test


def find_biggest_n_minus_one_product(A):

    # Builds suffix products.
    suffix_products = list(
        reversed(list(itertools.accumulate(reversed(A), operator.mul))))

    # Finds the biggest product of (n - 1) numbers.
    prefix_product, max_product = 1, float('-inf')
    for i in range(len(A)):
        suffix_product = suffix_products[i + 1] if i + 1 < len(A) else 1
        max_product = max(max_product, prefix_product * suffix_product)
        prefix_product *= A[i]
    return max_product


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main("max_product_all_but_one.py",
                                       'max_product_all_but_one.tsv',
                                       find_biggest_n_minus_one_product))

File: epi_judge_cpp_solutions/max_product_all_but_one.cc
#include <algorithm>
#include <functional>
#include <iterator>
#include <limits>
#include <numeric>
#include <vector>

#include "test_framework/generic_test.h"

using std::max;
using std::multiplies;
using std::numeric_limits;
using std::vector;

int FindBiggestNMinusOneProduct(const vector<int>& A) {
  // Builds suffix products.
  vector<int> suffix_products(size(A));
  partial_sum(crbegin(A), crend(A), rbegin(suffix_products), multiplies<int>());

  // Finds the biggest product of (n - 1) numbers.
  int prefix_product = 1, max_product = numeric_limits<int>::min();
  for (int i = 0; i < size(A); ++i) {
    int suffix_product = i + 1 < size(A) ? suffix_products[i + 1] : 1;
    max_product = max(max_product, prefix_product * suffix_product);
    prefix_product *= A[i];
  }
  return max_product;
}

int main(int argc, char* argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"A"};
  return GenericTestMain(
      args, "max_product_all_but_one.cc", "max_product_all_but_one.tsv",
      &FindBiggestNMinusOneProduct, DefaultComparator{}, param_names);
}
